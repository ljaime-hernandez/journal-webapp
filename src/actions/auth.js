import { firebase, googleAuthProvider } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import { types } from "../types/types";
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// 
export const startLoginEmailPassword = (email, password) => {

    // the dispatch contains functions from different reducers, but the useDispatch hook
    // is linked to redux, using the environment developed in the store file along with
    // our redux functions, the reducers are contained in the combineReducer method, allowing
    // us to dispatch several actions at once
    return (dispatch) => {

        dispatch(startLoading());

        // account authentication from regular email accounts, it returns a callback which we will
        // use to dispatch the login action from our auth reducer and, either it goes thru or not
        // will also dispatch the finishLoading function to enable the LoginScreen button for login
        firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {
            dispatch(finishLoading());
            dispatch(login(user.uid, user.displayName));
        }).catch( e => {
            dispatch(finishLoading());
            Swal.fire('Error', e.message,'error');
        })
    }
}

export const startGoogleLogin = () => {

    // as this function is a callback, the dispatch will be sent back to our authReducer
    // environment for it to handle user login status
    return (dispatch) => {

        // the googleAuthProvider is an authentication method returned from our firebase
        // db, along with the signInWithPopup the user will be able to use its google
        // original account to create an account in our webpage, and for us to
        // use the information for authentication and also render a proper user interface
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            })
        }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const logout = () => {
    return {
        type: types.logout
    }
}

export const startLogout = () => {
    return (dispatch) => {
        firebase.auth().signOut();
        dispatch(noteLogout());
        dispatch(logout());
    }
}
    
export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then( async ({user}) => {
            await user.updateProfile({displayName: name});

            dispatch(login(user.uid, user.displayName))
        }).catch( e => {
            console.log(e);
            // alert displaying an error message if theres errors on the user input
            // or if the email exists
            Swal.fire('Error', e.message,'error');
        })
    }
}

