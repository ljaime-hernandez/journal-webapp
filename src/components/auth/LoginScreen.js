import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

  const dispatch = useDispatch();
  const {ui} = useSelector(state => state);

  const {msgError, loading} = ui;


  const [values, handleInputChange] = useForm({
    email: 'email@email.com',
    password: '123456'
  })

  const {email, password} = values;

  const handleLogin = (e) => {
    e.preventDefault();

    if(isLoginValid()){
      dispatch(startLoginEmailPassword(email, password))
    }
  }

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  }

  const isLoginValid = () => {

    if ( !validator.isEmail(email)){
      dispatch(setError("The email input is incorrect"))
      return false;
    } else if( password < 5){
      dispatch(setError("Password should be longer than 4 characters"))
      return false;
    }

    dispatch(removeError())
    return true;
  }

  return (
  <>
      <h3 className="auth__title">Login</h3>

      <form 
        onSubmit={handleLogin}
        className='animate__animated animate__fadeIn'  
      >

      {
        msgError &&
        (<div className="auth__alert-error">
          {msgError}
        </div>)
      }

        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input" 
          value={password} 
          onChange={handleInputChange}     
        />

        <button
          type="submit"
          className="btn btn-primary btn-block"
          disabled={loading}
        >
          Login
        </button>

        {/*div containing copied button we will use to authorize an user with their google account */}
        <div className="auth__social-networks">
          <p>Login with social networks</p>

          <div 
              className="google-btn"
              onClick={handleGoogleLogin}
          >
              <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
              </div>
              <p className="btn-text">
                  <b>Sign in with google</b>
              </p>
          </div>

        </div>

        <Link 
          className="link"
          to="/auth/register"
        >
          Create new account
        </Link>
      </form>
  </>
  )
};
