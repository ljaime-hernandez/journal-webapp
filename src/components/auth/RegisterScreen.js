import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { setError, removeError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const {msgError} = useSelector( state => state.ui );

  const [values, handleInputChange] = useForm({
    name: 'Miguel',
    email: 'email@email.com',
    password: '123456',
    password2: '123456',
  })

  const {name, email, password, password2} = values;

  const handleRegister = (e) => {
    e.preventDefault();

    if(isFormValid()){
      dispatch(startRegisterWithEmailPasswordName(email, password, name))
    }
  }

  const isFormValid = () => {

    if(name.trim().length === 0){
      dispatch(setError("The name input is incorrect"))
      return false;
    } else if ( !validator.isEmail(email)){
      dispatch(setError("The email input is incorrect"))
      return false;
    } else if( password !== password2 || password2.length < 5){
      dispatch(setError("Password should be longer than 4 characters"))
      return false;
    }

    dispatch(removeError())
    return true;
  }

  return (
    <>
    <h3 className="auth__title">Register</h3>

    <form onSubmit={handleRegister}>

      {
        msgError &&
        (<div className="auth__alert-error">
          {msgError}
        </div>)
      }

    <input
        type="text"
        placeholder="Name"
        name="name"
        className="auth__input"
        autoComplete="off"
        value={name}
        onChange={handleInputChange}
        
      />

      <input
        type="text"
        placeholder="Email"
        name="email"
        className="auth__input"
        autoComplete="off"
        value={email}
        onChange={handleInputChange}
      />

      <input
        type="password"
        placeholder="Password"
        name="password"
        className="auth__input"
        value={password}
        onChange={handleInputChange}       
      />

<input
        type="password"
        placeholder="Confirm your Password"
        name="password2"
        className="auth__input"      
        value={password2}
        onChange={handleInputChange} 
      />

      <button
        type="submit"
        className="btn btn-primary btn-block mb-5"
      >
        Register
      </button>

      <Link 
        className="link"
        to="/auth/login"
      >
        Already registered?
      </Link>
    </form>
</>
  )
}
