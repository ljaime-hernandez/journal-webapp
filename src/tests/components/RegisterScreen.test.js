import '@testing-library/jest-dom';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { RegisterScreen } from "../../components/auth/RegisterScreen";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};
let store = mockStore(initialState);

describe('tests on RegisterScreen component', () => {


    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen/>
            </MemoryRouter>
        </Provider>
        );

    beforeEach(()=> {
        store = mockStore(initialState);
        jest.clearAllMocks();
    });

    test('should render properly', () => {
        
        expect(wrapper).toMatchSnapshot();
    });

    test('should dispatch actions accordingly', () => {
      
        const emailField = wrapper.find('input[name="email"]');

        emailField.simulate('change', {
            target: {
                value: '',
                name: 'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();
        console.log(actions);

    });
    
    
})