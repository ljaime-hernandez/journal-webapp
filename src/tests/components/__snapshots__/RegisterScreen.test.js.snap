// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on RegisterScreen component should render properly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Miguel"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="email@email.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm your Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
